pkg/sentry/server/validator/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,241 @@
+		if _, err := cache.GetInformer(ctx, obj); err != nil {
+			return nil, err
+		}
+	}
+
+	return &kubernetes{
+		auth:           kubeClient.AuthenticationV1(),
+		client:         cache,
+		ready:          cache.WaitForCacheSync,
+		sentryAudience: opts.SentryID.String(),
+		controlPlaneNS: opts.ControlPlaneNS,
+		controlPlaneTD: opts.SentryID.TrustDomain(),
+	}, nil
+}
+
+func (k *kubernetes) Start(ctx context.Context) error {
+	if err := k.client.(cache.Cache).Start(ctx); err != nil {
+		return err
+	}
+
+	<-ctx.Done()
+	return nil
+}
+
+func (k *kubernetes) Validate(ctx context.Context, req *sentryv1pb.SignCertificateRequest) (spiffeid.TrustDomain, error) {
+	if !k.ready(ctx) {
+		return spiffeid.TrustDomain{}, errors.New("validator not ready")
+	}
+
+	// The TrustDomain field is ignored by the Kubernetes validator.
+	if _, err := internal.Validate(ctx, req); err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	prts, err := k.executeTokenReview(ctx, req.GetToken(), LegacyServiceAccountAudience, k.sentryAudience)
+	if err != nil {
+		return spiffeid.TrustDomain{}, err
+	}
+
+	if len(prts) != 4 || prts[0] != "system" {
+		return spiffeid.TrustDomain{}, errors.New("provided token is not a properly structured service account token")
+	}
+
+	if prts[2] != req.Namespace {
+		return spiffeid.TrustDomain{}, fmt.Errorf("namespace mismatch; received namespace: %s", req.Namespace)
+	}
+
+	// We have already validated to the token against Kubernetes API server, so
+	// we do not need to supply a key.
+	ptoken, err := jwt.ParseInsecure([]byte(req.GetToken()), jwt.WithTypedClaim("kubernetes.io", new(k8sClaims)))
+	if err != nil {
+		return spiffeid.TrustDomain{}, fmt.Errorf("failed to parse Kubernetes token: %s", err)
+	}
+	claimsT, ok := ptoken.Get("kubernetes.io")
+	if !ok {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+	claims, ok := claimsT.(*k8sClaims)
+	if !ok || len(claims.Pod.Name) == 0 {
+		return spiffeid.TrustDomain{}, errMissingPodClaim
+	}
+
+	var pod corev1.Pod
+	err = k.client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: claims.Pod.Name}, &pod)
+	if err != nil {
+		log.Errorf("Failed to get pod %s/%s for requested identity: %s", req.Namespace, claims.Pod.Name, err)
+		return spiffeid.TrustDomain{}, errors.New("failed to get pod of identity")
+	}
+	expID, ok := pod.GetAnnotations()[annotations.KeyAppID]
+	if !ok {
+		expID = pod.GetName()
+	}
+
+	if pod.Spec.ServiceAccountName != prts[3] {
+		log.Errorf("Service account on pod %s/%s does not match token", req.Namespace, claims.Pod.Name)
+		return spiffeid.TrustDomain{}, errors.New("pod service account mismatch")
+	}
+
+	// TODO: @joshvanl: Before v1.11, the injector instructed daprd to request
+	// for the ID containing their namespace and service account (ns:sa). This
+	// is wrong- dapr identities are based on daprd namespace + _app ID_.
+	// Remove this allowance in v1.12.
+	if req.Namespace+":"+pod.Spec.ServiceAccountName == req.Id {
+		req.Id = expID
```

Comment:
But you validated the req.ID is < 64 characters before this even though it wasn't the app ID.  Which subsequently seems to make 1.12 unusable if you have any namespaces + app ID's of much length at ll.

Reply:
Thanks @a-shoemaker, indeed this looks like a bug in 1.12 which needs to be fixed.

Reply:
@a-shoemaker I have opened up an issue [here](https://github.com/dapr/dapr/issues/7079) to track.